src/main/java/at/hannibal2/skyhanni/features/mining/glacitemineshaft/MineshaftWaypoints.kt

```diff
@@ -0,0 +1,88 @@
+package at.hannibal2.skyhanni.features.mining.glacitemineshaft
+
+import at.hannibal2.skyhanni.SkyHanniMod
+import at.hannibal2.skyhanni.data.IslandType
+import at.hannibal2.skyhanni.data.PartyAPI
+import at.hannibal2.skyhanni.events.IslandChangeEvent
+import at.hannibal2.skyhanni.events.LorenzKeyPressEvent
+import at.hannibal2.skyhanni.events.LorenzRenderWorldEvent
+import at.hannibal2.skyhanni.events.LorenzWorldChangeEvent
+import at.hannibal2.skyhanni.test.GriffinUtils.drawWaypointFilled
+import at.hannibal2.skyhanni.utils.ChatUtils
+import at.hannibal2.skyhanni.utils.LocationUtils
+import at.hannibal2.skyhanni.utils.LocationUtils.distanceToPlayer
+import at.hannibal2.skyhanni.utils.LorenzUtils.isInIsland
+import at.hannibal2.skyhanni.utils.RenderUtils.drawDynamicText
+import at.hannibal2.skyhanni.utils.SimpleTimeMark
+import net.minecraft.client.Minecraft
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import kotlin.time.Duration.Companion.milliseconds
+
+object MineshaftWaypoints {
+    private val config get() = SkyHanniMod.feature.mining.mineshaftWaypoints
+
+    @Suppress("ConstPropertyName")
+    private const val blocksForward: Int = 7
+
+    val waypoints: MutableList<Waypoint> = mutableListOf()
+    private var timeLastShared = SimpleTimeMark.farPast()
+
+    @Suppress("UNUSED_PARAMETER")
+    @SubscribeEvent
+    fun onWorldChange(event: LorenzWorldChangeEvent) {
+        waypoints.clear()
+    }
+
+    @SubscribeEvent
+    fun onIslandChange(event: IslandChangeEvent) {
+        if (event.newIsland != IslandType.MINESHAFT) return
+
+        val playerLocation = LocationUtils.playerLocation().round(0).add(y = -1)
+
+        if (config.entranceLocation) {
+            val waypointType = WaypointsType.ENTRANCE
+            waypoints.add(Waypoint(waypointType = waypointType, location = playerLocation, shared = true))
+        }
+
+        if (config.ladderLocation) {
+            val waypointType = WaypointsType.LADDER
+            val vec = Minecraft.getMinecraft().thePlayer.horizontalFacing.directionVec
+            val location = playerLocation
+                // Move 6 blocks in front of the player to be in the ladder shaft
+                .add(x = vec.x * blocksForward, y = -15, z = vec.z * blocksForward)
+                // Adjust 2 blocks to the right to be in the center of the ladder shaft
+                .add(x = vec.z * -2, z = vec.x * 2)
```

Comment:
should probably make sure math is correct, but with the 10 mineshafts I've joined it seemed to be correct.